File: leetcode/816_Ambiguous_Coordinates/816_Ambiguous_Coordinates.cpp
class Solution {
public:
    bool validate(string l, string r) {
        if (l.length() > 1 && l[0] == '0') return false;
        if (r.length() > 0 && r[r.length() - 1] == '0') return false;
        return true;
    }
    void backTracking(vector<string>& strList, vector<string>& res, string tempStr, int index) {
        if (index == 2) {
            res.push_back(tempStr);
            return;
        }
        string str = strList[index];
            for (int i = 1; i <= str.length(); i++) {
                if (validate(str.substr(0, i), str.substr(i))) {
                    string temp = tempStr;
                    temp += str.substr(0, i) + (i == str.length() ? "" : ".") + str.substr(i);
                    if (index == 0) temp = "(" + temp + ", ";
                    else temp += ")";
                    backTracking(strList, res, temp, index + 1);
                }
            }
    }
    
    vector<string> ambiguousCoordinates(string S) {
        S = S.substr(1, S.length() - 2);
        vector<string> res = vector<string>();
        for (int i = 1; i < S.length(); i++) {
            vector<string> temp = {S.substr(0, i), S.substr(i)};
            backTracking(temp, res, string(), 0);        
        }
        
        return res;
    }
};

File: leetcode/816_Ambiguous_Coordinates/816_Ambiguous_Coordinates.java
class Solution {
    private boolean validate(String l, String r) {
        if (l.charAt(0) == '0' && l.length() > 1) return false;
        if (r!= null && r.length() > 0 && r.charAt(r.length() - 1) == '0') return false;
        return true;
    }
    
    private void backTracking(List<String> strList, List<String> res, String tempStr, int index) {
        if (index == 2) {
            res.add(tempStr);
            return;
        }
        String str = strList.get(index);
        for (int i = 1; i <= str.length(); i++) {
            if (validate(str.substring(0, i), str.substring(i))) {
                String temp = tempStr;
                temp += str.substring(0, i) + (i == str.length() ? "" : ".") + str.substring(i);
                if (index == 0) temp = "(" + temp + ", ";
                else temp += ")";
                backTracking(strList, res, temp, index + 1);
            }
        }
    }
    
    public List<String> ambiguousCoordinates(String S) {
        S = S.substring(1, S.length() - 1);
        List<String> res = new ArrayList<String>();
        for (int i = 1; i < S.length(); i++) {
            String l = S.substring(0, i);
            String r = S.substring(i);
            backTracking(Arrays.asList(l, r), res, new String(), 0);
        }
        return res;
    }
}
